server/schedule/labeler/rules.go

```diff
@@ -68,6 +73,69 @@ type LabelRulePatch struct {
 	DeleteRules []string     `json:"deletes"`
 }
 
+func (l *RegionLabel) expireBefore(t time.Time) bool {
+	return l.getExpire().Before(t)
+}
+
+var unlimittedExpire time.Time
+
+func init() {
+	unlimittedExpire, _ = time.Parse(time.UnixDate, "Mon Jan 12 22:22:22 MST 2222")
+}
+
+func (l *RegionLabel) checkAndAdjustExpire() (err error) {
+	if len(l.TTL) == 0 {
+		l.expire = unlimittedExpire
+		return
+	}
+	ttl, err := time.ParseDuration(l.TTL)
+	if err != nil {
+		return err
+	}
+	var startAt time.Time
+	if len(l.StartAt) == 0 {
+		startAt = time.Now()
+		l.StartAt = startAt.Format(time.UnixDate)
+	} else {
+		startAt, err = time.Parse(time.UnixDate, l.StartAt)
+		if err != nil {
+			return err
+		}
+	}
+	l.expire = startAt.Add(ttl)
+	return nil
+}
+
+// GetExpire returns time the rule expired.
+func (l *RegionLabel) getExpire() time.Time {
+	if !l.expire.IsZero() {
+		return l.expire
+	}
+	l.checkAndAdjustExpire()
+	return l.expire
+}
+
+func (rule *LabelRule) checkAndRemoveExpireLabels(now time.Time) bool {
+	minExpire := unlimittedExpire
+	labels := make([]RegionLabel, 0)
+	for _, l := range rule.Labels {
+		if l.expireBefore(now) {
+			continue
+		}
+		labels = append(labels, l)
+		if l.expireBefore(minExpire) {
+			minExpire = l.expire
+		}
+	}
+
+	if len(labels) == len(rule.Labels) && rule.minExpire.Equal(minExpire) {
```

Comment:
Why not directly compare with `rule.minExpire` for each label?

Reply:
Good catch